File: backend/src/recommendation/explainability.ts
import { predictEnsembleProbabilities, type TrainedEnsembleModels } from "./models.js";
import type {
  CertificationSignalKey,
  CareerPathKey,
  CompetencyKey,
  ExplainabilityFactor,
  ExplainabilityMethodReport,
  RecommendationExplainability,
} from "./types.js";
import { clamp01, SeededRng } from "./utils.js";

type LabelMap = Record<CompetencyKey, string>;

type ExplainerInput = {
  models: TrainedEnsembleModels;
  featureVector: number[];
  featureKeys: CompetencyKey[];
  labels: LabelMap;
  careerIndex: number;
  careerName: string;
  pathKey: CareerPathKey;
  methodPreference?: "auto" | "shap" | "lime";
  additionalFactors?: Array<{
    key: CertificationSignalKey;
    label: string;
    value: number;
    contribution: number;
  }>;
};

function scoreCareer(models: TrainedEnsembleModels, featureVector: number[], careerIndex: number) {
  const probabilities = predictEnsembleProbabilities(models, featureVector);
  return clamp01(probabilities.ensemble[careerIndex] ?? 0);
}

function solveLinearSystem(matrix: number[][], vector: number[]) {
  const n = vector.length;
  const a = matrix.map((row) => [...row]);
  const b = [...vector];

  for (let col = 0; col < n; col += 1) {
    let pivot = col;
    for (let row = col + 1; row < n; row += 1) {
      if (Math.abs(a[row][col]) > Math.abs(a[pivot][col])) {
        pivot = row;
      }
    }

    if (Math.abs(a[pivot][col]) < 1e-10) {
      return Array.from({ length: n }, () => 0);
    }

    if (pivot !== col) {
      const tempRow = a[col];
      a[col] = a[pivot];
      a[pivot] = tempRow;

      const tempB = b[col];
      b[col] = b[pivot];
      b[pivot] = tempB;
    }

    const pivotValue = a[col][col];
    for (let j = col; j < n; j += 1) {
      a[col][j] /= pivotValue;
    }
    b[col] /= pivotValue;

    for (let row = 0; row < n; row += 1) {
      if (row === col) continue;
      const factor = a[row][col];
      if (Math.abs(factor) < 1e-12) continue;
      for (let j = col; j < n; j += 1) {
        a[row][j] -= factor * a[col][j];
      }
      b[row] -= factor * b[col];
    }
  }

  return b;
}

function buildFactors(
  keys: CompetencyKey[],
  labels: LabelMap,
  featureVector: number[],
  contributions: number[]
) {
  const totalImpact = contributions.reduce((sum, value) => sum + Math.abs(value), 0) || 1;
  const factors: ExplainabilityFactor[] = keys.map((key, index) => ({
    key,
    label: labels[key],
    value: featureVector[index] ?? 0,
    contribution: contributions[index] ?? 0,
    impactPct: (Math.abs(contributions[index] ?? 0) / totalImpact) * 100,
    direction: (contributions[index] ?? 0) >= 0 ? "positive" : "negative",
  }));

  return factors.sort((left, right) => Math.abs(right.contribution) - Math.abs(left.contribution));
}

function mergeAndNormalizeFactors(
  baseFactors: ExplainabilityFactor[],
  additionalFactors: Array<{
    key: CertificationSignalKey;
    label: string;
    value: number;
    contribution: number;
  }>
) {
  const merged: ExplainabilityFactor[] = [
    ...baseFactors.map((factor) => ({ ...factor, source: "competency" as const })),
    ...additionalFactors
      .filter((factor) => Math.abs(factor.contribution) > 1e-8)
      .map((factor) => ({
        key: factor.key,
        label: factor.label,
        value: factor.value,
        contribution: factor.contribution,
        impactPct: 0,
        direction: factor.contribution >= 0 ? ("positive" as const) : ("negative" as const),
        source: "certification" as const,
      })),
  ];

  const totalImpact = merged.reduce((sum, factor) => sum + Math.abs(factor.contribution), 0) || 1;
  const normalized = merged.map((factor) => ({
    ...factor,
    impactPct: (Math.abs(factor.contribution) / totalImpact) * 100,
  }));

  return normalized.sort((left, right) => Math.abs(right.contribution) - Math.abs(left.contribution));
}

function formatImpactPct(value: number) {
  if (!Number.isFinite(value) || value <= 0) {
    return "0%";
  }
  if (value >= 10) {
    return `${Math.round(value)}%`;
  }
  if (value >= 1) {
    return `${value.toFixed(1).replace(/\.0$/, "")}%`;
  }
  return `${value.toFixed(2).replace(/0+$/, "").replace(/\.$/, "")}%`;
}

function toNaturalList(items: string[]) {
  if (items.length === 0) return "";
  if (items.length === 1) return items[0];
  if (items.length === 2) return `${items[0]} and ${items[1]}`;
  return `${items.slice(0, -1).join(", ")}, and ${items[items.length - 1]}`;
}

function buildNarrative(careerName: string, factors: ExplainabilityFactor[]) {
  const topPositive = factors
    .filter((factor) => factor.contribution > 0)
    .sort((left, right) => right.contribution - left.contribution)
    .slice(0, 3);
  const topNegative = factors
    .filter((factor) => factor.contribution < 0)
    .sort((left, right) => left.contribution - right.contribution)[0];

  if (topPositive.length === 0) {
    if (topNegative) {
      return `We recommended ${careerName} because your overall profile aligns best with this role. The main factor reducing your score is ${topNegative.label}.`;
    }
    return `We recommended ${careerName} because your overall profile aligns best with this role.`;
  }

  const highlights = topPositive.map(
    (factor) => `${factor.label} (${formatImpactPct(factor.impactPct)})`
  );
  const topSignals = toNaturalList(highlights);

  let narrative = `We recommended ${careerName} because your strongest signals were ${topSignals}.`;
  narrative += " These signals contributed most to your match for this role.";
  if (topNegative) {
    narrative += ` The main area lowering your score is ${topNegative.label}.`;
  }
  return narrative;
}

function createMethodReport(args: {
  method: "shap" | "lime";
  careerName: string;
  pathKey: CareerPathKey;
  featureVector: number[];
  keys: CompetencyKey[];
  labels: LabelMap;
  contributions: number[];
  baseScore: number;
  predictedScore: number;
  reconstructedScore: number;
  fidelity: number;
  runtimeMs: number;
  additionalFactors?: Array<{
    key: CertificationSignalKey;
    label: string;
    value: number;
    contribution: number;
  }>;
}): ExplainabilityMethodReport {
  const baseFactors = buildFactors(args.keys, args.labels, args.featureVector, args.contributions);
  const factors = mergeAndNormalizeFactors(baseFactors, args.additionalFactors ?? []);
  return {
    method: args.method,
    careerName: args.careerName,
    pathKey: args.pathKey,
    baseScore: clamp01(args.baseScore),
    predictedScore: clamp01(args.predictedScore),
    reconstructedScore: clamp01(args.reconstructedScore),
    narrative: buildNarrative(args.careerName, factors),
    quality: {
      runtimeMs: args.runtimeMs,
      fidelity: clamp01(args.fidelity),
    },
    factors,
  };
}

function explainWithShap(
  models: TrainedEnsembleModels,
  featureVector: number[],
  careerIndex: number
) {
  const start = Date.now();
  const rng = new SeededRng(20260301 + careerIndex);
  const featureCount = featureVector.length;
  const means = models.featureStats?.means ?? Array.from({ length: featureCount }, () => 0.5);
  const permutations = Math.max(120, featureCount * 20);

  const contributions = Array.from({ length: featureCount }, () => 0);
  const baseline = means.map((value) => clamp01(value));
  const baseScore = scoreCareer(models, baseline, careerIndex);
  const predictedScore = scoreCareer(models, featureVector, careerIndex);

  for (let round = 0; round < permutations; round += 1) {
    const indices = Array.from({ length: featureCount }, (_, index) => index);
    rng.shuffle(indices);

    const current = [...baseline];
    let prevScore = scoreCareer(models, current, careerIndex);

    for (const index of indices) {
      current[index] = featureVector[index];
      const nextScore = scoreCareer(models, current, careerIndex);
      contributions[index] += nextScore - prevScore;
      prevScore = nextScore;
    }
  }

  for (let i = 0; i < contributions.length; i += 1) {
    contributions[i] /= permutations;
  }

  const reconstructed = baseScore + contributions.reduce((sum, value) => sum + value, 0);
  const error = Math.abs(reconstructed - predictedScore);
  const fidelity = clamp01(1 - error / 0.08);

  return {
    contributions,
    baseScore,
    predictedScore,
    reconstructedScore: reconstructed,
    fidelity,
    runtimeMs: Date.now() - start,
  };
}

function explainWithLime(
  models: TrainedEnsembleModels,
  featureVector: number[],
  careerIndex: number
) {
  const start = Date.now();
  const rng = new SeededRng(20260317 + careerIndex);
  const featureCount = featureVector.length;
  const means = models.featureStats?.means ?? Array.from({ length: featureCount }, () => 0.5);
  const stds = models.featureStats?.stds ?? Array.from({ length: featureCount }, () => 0.2);

  const sampleCount = 280;
  const kernelWidth = 0.9;
  const lambda = 1e-3;

  const samples: number[][] = [];
  const targets: number[] = [];
  const weights: number[] = [];

  for (let i = 0; i < sampleCount; i += 1) {
    const sample = featureVector.map((value, featureIndex) => {
      if (i === 0) return value;
      const keepOriginal = rng.next() > 0.18;
      if (!keepOriginal) {
        return means[featureIndex];
      }

      const noise = (rng.next() - 0.5) * 1.2 * (stds[featureIndex] || 0.2);
      return clamp01(value + noise);
    });

    const distance = Math.sqrt(
      sample.reduce((sum, value, featureIndex) => {
        const denom = Math.max(stds[featureIndex] || 0.2, 1e-3);
        const delta = (value - featureVector[featureIndex]) / denom;
        return sum + delta * delta;
      }, 0) / Math.max(1, featureCount)
    );
    const weight = Math.exp(-(distance * distance) / (kernelWidth * kernelWidth));

    samples.push(sample);
    targets.push(scoreCareer(models, sample, careerIndex));
    weights.push(weight);
  }

  const dim = featureCount + 1;
  const a = Array.from({ length: dim }, () => Array.from({ length: dim }, () => 0));
  const b = Array.from({ length: dim }, () => 0);

  for (let row = 0; row < samples.length; row += 1) {
    const x = [1, ...samples[row]];
    const y = targets[row];
    const w = weights[row];

    for (let i = 0; i < dim; i += 1) {
      b[i] += w * x[i] * y;
      for (let j = 0; j < dim; j += 1) {
        a[i][j] += w * x[i] * x[j];
      }
    }
  }

  for (let i = 0; i < dim; i += 1) {
    a[i][i] += lambda;
  }

  const beta = solveLinearSystem(a, b);
  const intercept = beta[0] ?? 0;
  const coefs = beta.slice(1);

  const baseScore = scoreCareer(models, means, careerIndex);
  const contributions = coefs.map(
    (coef, featureIndex) => coef * (featureVector[featureIndex] - means[featureIndex])
  );
  const predictedScore = scoreCareer(models, featureVector, careerIndex);
  const reconstructedScore = intercept + coefs.reduce((sum, coef, index) => sum + coef * featureVector[index], 0);

  const weightedTotal = weights.reduce((sum, value) => sum + value, 0) || 1;
  const weightedMeanY =
    targets.reduce((sum, value, index) => sum + value * weights[index], 0) / weightedTotal;
  const sse = targets.reduce((sum, value, index) => {
    const predicted =
      intercept + coefs.reduce((acc, coef, featureIndex) => acc + coef * samples[index][featureIndex], 0);
    const err = value - predicted;
    return sum + weights[index] * err * err;
  }, 0);
  const sst = targets.reduce((sum, value, index) => {
    const err = value - weightedMeanY;
    return sum + weights[index] * err * err;
  }, 0);
  const fidelity = sst > 1e-12 ? clamp01(1 - sse / sst) : 0.5;

  const centeredContrib = contributions;
  const reconstructedFromCentered = baseScore + centeredContrib.reduce((sum, value) => sum + value, 0);

  return {
    contributions: centeredContrib,
    baseScore,
    predictedScore,
    reconstructedScore: reconstructedFromCentered || reconstructedScore,
    fidelity,
    runtimeMs: Date.now() - start,
  };
}

export function buildCareerExplainability({
  models,
  featureVector,
  featureKeys,
  labels,
  careerIndex,
  careerName,
  pathKey,
  methodPreference = "auto",
  additionalFactors = [],
}: ExplainerInput): RecommendationExplainability {
  const shap = explainWithShap(models, featureVector, careerIndex);
  const lime = explainWithLime(models, featureVector, careerIndex);

  const shapReport = createMethodReport({
    method: "shap",
    careerName,
    pathKey,
    featureVector,
    keys: featureKeys,
    labels,
    contributions: shap.contributions,
    baseScore: shap.baseScore,
    predictedScore: shap.predictedScore,
    reconstructedScore: shap.reconstructedScore,
    fidelity: shap.fidelity,
    runtimeMs: shap.runtimeMs,
    additionalFactors: additionalFactors ?? [],
  });
  const limeReport = createMethodReport({
    method: "lime",
    careerName,
    pathKey,
    featureVector,
    keys: featureKeys,
    labels,
    contributions: lime.contributions,
    baseScore: lime.baseScore,
    predictedScore: lime.predictedScore,
    reconstructedScore: lime.reconstructedScore,
    fidelity: lime.fidelity,
    runtimeMs: lime.runtimeMs,
    additionalFactors: additionalFactors ?? [],
  });

  const shapScore = shapReport.quality.fidelity - shapReport.quality.runtimeMs / 15000;
  const limeScore = limeReport.quality.fidelity - limeReport.quality.runtimeMs / 15000;

  let selectedMethod: "shap" | "lime";
  if (methodPreference === "shap" || methodPreference === "lime") {
    selectedMethod = methodPreference;
  } else {
    selectedMethod = shapScore >= limeScore - 0.01 ? "shap" : "lime";
  }

  const reason =
    selectedMethod === "shap"
      ? "SHAP was selected because it provides additive, more stable contributions while keeping comparable or better fidelity."
      : "LIME was selected because it achieved higher local surrogate fidelity for this prediction.";

  return {
    selectedMethod,
    reason,
    topCareer: selectedMethod === "shap" ? shapReport : limeReport,
    comparison: {
      shap: shapReport,
      lime: limeReport,
    },
  };
}

File: Python/app/explainability.py
from __future__ import annotations

import math
import random
import time
from typing import Any, Callable, Dict, List


def clamp01(value: float) -> float:
    return max(0.0, min(1.0, value))


def solve_linear_system(matrix: List[List[float]], vector: List[float]) -> List[float]:
    n = len(vector)
    a = [row[:] for row in matrix]
    b = vector[:]

    for col in range(n):
      pivot = col
      for row in range(col + 1, n):
          if abs(a[row][col]) > abs(a[pivot][col]):
              pivot = row

      if abs(a[pivot][col]) < 1e-10:
          return [0.0 for _ in range(n)]

      if pivot != col:
          a[col], a[pivot] = a[pivot], a[col]
          b[col], b[pivot] = b[pivot], b[col]

      pivot_value = a[col][col]
      for j in range(col, n):
          a[col][j] /= pivot_value
      b[col] /= pivot_value

      for row in range(n):
          if row == col:
              continue
          factor = a[row][col]
          if abs(factor) < 1e-12:
              continue
          for j in range(col, n):
              a[row][j] -= factor * a[col][j]
          b[row] -= factor * b[col]

    return b


def build_factors(
    keys: List[str],
    labels: Dict[str, str],
    feature_vector: List[float],
    contributions: List[float],
    relevant_keys: List[str] | None = None,
) -> List[Dict[str, Any]]:
    relevant_lookup = set(relevant_keys or [])
    filtered_entries = [
        (index, key)
        for index, key in enumerate(keys)
        if not relevant_lookup or key in relevant_lookup
    ]
    total_impact = (
        sum(
            abs(contributions[index] if index < len(contributions) else 0.0)
            for index, _ in filtered_entries
        )
        or 1.0
    )
    factors = [
        {
            "key": key,
            "label": labels[key],
            "value": feature_vector[index] if index < len(feature_vector) else 0.0,
            "contribution": contributions[index] if index < len(contributions) else 0.0,
            "impactPct": abs(contributions[index] if index < len(contributions) else 0.0) / total_impact * 100,
            "direction": "positive" if (contributions[index] if index < len(contributions) else 0.0) >= 0 else "negative",
        }
        for index, key in filtered_entries
    ]
    return sorted(factors, key=lambda item: abs(item["contribution"]), reverse=True)


def merge_and_normalize_factors(
    base_factors: List[Dict[str, Any]],
    additional_factors: List[Dict[str, Any]],
) -> List[Dict[str, Any]]:
    merged = [
        {**factor, "source": "competency"} for factor in base_factors
    ] + [
        {
            "key": factor["key"],
            "label": factor["label"],
            "value": factor["value"],
            "contribution": factor["contribution"],
            "impactPct": 0.0,
            "direction": "positive" if factor["contribution"] >= 0 else "negative",
            "source": "certification",
        }
        for factor in additional_factors
        if abs(factor["contribution"]) > 1e-8
    ]

    total_impact = sum(abs(factor["contribution"]) for factor in merged) or 1.0
    normalized = [
        {**factor, "impactPct": abs(factor["contribution"]) / total_impact * 100}
        for factor in merged
    ]
    return sorted(normalized, key=lambda item: abs(item["contribution"]), reverse=True)


def format_impact_pct(value: float) -> str:
    if not math.isfinite(value) or value <= 0:
        return "0%"
    if value >= 10:
        return f"{round(value)}%"
    if value >= 1:
        return f"{value:.1f}".rstrip("0").rstrip(".") + "%"
    return f"{value:.2f}".rstrip("0").rstrip(".") + "%"


def to_natural_list(items: List[str]) -> str:
    if not items:
        return ""
    if len(items) == 1:
        return items[0]
    if len(items) == 2:
        return f"{items[0]} and {items[1]}"
    return f"{', '.join(items[:-1])}, and {items[-1]}"


def build_narrative(
    career_name: str,
    factors: List[Dict[str, Any]],
    gap_recommendations: Dict[str, str],
) -> str:
    top_positive = [factor for factor in factors if factor["contribution"] > 0][:3]
    top_negative = next((factor for factor in factors if factor["contribution"] < 0), None)
    top_improvement_areas = [
        factor
        for factor in factors
        if factor.get("source") != "certification"
        and factor["contribution"] < 0
        and factor.get("key") in gap_recommendations
    ][:2]

    if not top_positive:
        if top_negative:
            narrative = (
                f"We recommended {career_name} because your overall profile aligns best with this role. "
                f"The main factor reducing your score is {top_negative['label']}."
            )
        else:
            narrative = f"We recommended {career_name} because your overall profile aligns best with this role."
        if top_improvement_areas:
            guidance = " ".join(
                f"Improve {factor['label']} by {gap_recommendations[str(factor['key'])].lower()}"
                for factor in top_improvement_areas
            )
            narrative += f" To strengthen your fit further, {guidance}."
        return narrative

    highlights = [
        f"{factor['label']} ({format_impact_pct(float(factor['impactPct']))})"
        for factor in top_positive
    ]
    top_signals = to_natural_list(highlights)
    narrative = (
        f"We recommended {career_name} because your strongest signals were {top_signals}. "
        "These signals contributed most to your match for this role."
    )
    if top_negative:
        narrative += f" The main area lowering your score is {top_negative['label']}."
    if top_improvement_areas:
        areas = to_natural_list([factor["label"] for factor in top_improvement_areas])
        guidance = " ".join(
            f"For {factor['label']}, {gap_recommendations[str(factor['key'])].lower()}"
            for factor in top_improvement_areas
        )
        narrative += f" The main areas to improve next are {areas}. {guidance}."
    return narrative


def create_method_report(
    *,
    method: str,
    career_name: str,
    path_key: str,
    feature_vector: List[float],
    keys: List[str],
    labels: Dict[str, str],
    contributions: List[float],
    base_score: float,
    predicted_score: float,
    reconstructed_score: float,
    fidelity: float,
    runtime_ms: int,
    additional_factors: List[Dict[str, Any]],
    gap_recommendations: Dict[str, str],
    relevant_keys: List[str] | None = None,
) -> Dict[str, Any]:
    base_factors = build_factors(
        keys,
        labels,
        feature_vector,
        contributions,
        relevant_keys=relevant_keys,
    )
    factors = merge_and_normalize_factors(base_factors, additional_factors)
    return {
        "method": method,
        "careerName": career_name,
        "pathKey": path_key,
        "baseScore": clamp01(base_score),
        "predictedScore": clamp01(predicted_score),
        "reconstructedScore": clamp01(reconstructed_score),
        "narrative": build_narrative(career_name, factors, gap_recommendations),
        "quality": {
            "runtimeMs": runtime_ms,
            "fidelity": clamp01(fidelity),
        },
        "factors": factors,
    }


def create_empty_method_report(
    *,
    method: str,
    career_name: str,
    path_key: str,
) -> Dict[str, Any]:
    return {
        "method": method,
        "careerName": career_name,
        "pathKey": path_key,
        "baseScore": 0.0,
        "predictedScore": 0.0,
        "reconstructedScore": 0.0,
        "narrative": "",
        "quality": {
            "runtimeMs": 0,
            "fidelity": 0.0,
        },
        "factors": [],
    }


def explain_with_shap(
    predict_score: Callable[[List[float], int], float],
    feature_vector: List[float],
    career_index: int,
    means: List[float],
) -> Dict[str, Any]:
    start = time.time()
    rng = random.Random(20260301 + career_index)
    feature_count = len(feature_vector)
    permutations = max(120, feature_count * 20)

    contributions = [0.0 for _ in range(feature_count)]
    baseline = [clamp01(value) for value in means]
    base_score = predict_score(baseline, career_index)
    predicted_score = predict_score(feature_vector, career_index)

    for _ in range(permutations):
        indices = list(range(feature_count))
        rng.shuffle(indices)
        current = baseline[:]
        prev_score = predict_score(current, career_index)
        for index in indices:
            current[index] = feature_vector[index]
            next_score = predict_score(current, career_index)
            contributions[index] += next_score - prev_score
            prev_score = next_score

    contributions = [value / permutations for value in contributions]
    reconstructed = base_score + sum(contributions)
    error = abs(reconstructed - predicted_score)
    fidelity = clamp01(1 - error / 0.08)

    return {
        "contributions": contributions,
        "baseScore": base_score,
        "predictedScore": predicted_score,
        "reconstructedScore": reconstructed,
        "fidelity": fidelity,
        "runtimeMs": int((time.time() - start) * 1000),
    }


def explain_with_lime(
    predict_score: Callable[[List[float], int], float],
    feature_vector: List[float],
    career_index: int,
    means: List[float],
    stds: List[float],
) -> Dict[str, Any]:
    start = time.time()
    rng = random.Random(20260317 + career_index)
    feature_count = len(feature_vector)
    sample_count = 280
    kernel_width = 0.9
    lambda_reg = 1e-3

    samples: List[List[float]] = []
    targets: List[float] = []
    weights: List[float] = []

    for i in range(sample_count):
        sample: List[float] = []
        for feature_index, value in enumerate(feature_vector):
            if i == 0:
                sample.append(value)
                continue
            keep_original = rng.random() > 0.18
            if not keep_original:
                sample.append(means[feature_index])
                continue
            noise = (rng.random() - 0.5) * 1.2 * (stds[feature_index] or 0.2)
            sample.append(clamp01(value + noise))

        distance = math.sqrt(
            sum(
                (((sample[idx] - feature_vector[idx]) / max(stds[idx] or 0.2, 1e-3)) ** 2)
                for idx in range(feature_count)
            )
            / max(1, feature_count)
        )
        weight = math.exp(-(distance * distance) / (kernel_width * kernel_width))
        samples.append(sample)
        targets.append(predict_score(sample, career_index))
        weights.append(weight)

    dim = feature_count + 1
    a = [[0.0 for _ in range(dim)] for _ in range(dim)]
    b = [0.0 for _ in range(dim)]

    for row, sample in enumerate(samples):
        x = [1.0, *sample]
        y = targets[row]
        w = weights[row]
        for i in range(dim):
            b[i] += w * x[i] * y
            for j in range(dim):
                a[i][j] += w * x[i] * x[j]

    for i in range(dim):
        a[i][i] += lambda_reg

    beta = solve_linear_system(a, b)
    intercept = beta[0] if beta else 0.0
    coefs = beta[1:]
    base_score = predict_score(means, career_index)
    contributions = [
        (coefs[idx] if idx < len(coefs) else 0.0) * (feature_vector[idx] - means[idx])
        for idx in range(feature_count)
    ]
    predicted_score = predict_score(feature_vector, career_index)
    reconstructed_score = intercept + sum(
        (coefs[idx] if idx < len(coefs) else 0.0) * feature_vector[idx]
        for idx in range(feature_count)
    )

    weighted_total = sum(weights) or 1.0
    weighted_mean_y = sum(targets[idx] * weights[idx] for idx in range(len(targets))) / weighted_total
    sse = 0.0
    sst = 0.0
    for idx, sample in enumerate(samples):
        predicted = intercept + sum(
            (coefs[feature_index] if feature_index < len(coefs) else 0.0) * sample[feature_index]
            for feature_index in range(feature_count)
        )
        err = targets[idx] - predicted
        sse += weights[idx] * err * err
        centered = targets[idx] - weighted_mean_y
        sst += weights[idx] * centered * centered
    fidelity = clamp01(1 - sse / sst) if sst > 1e-12 else 0.5

    reconstructed_from_centered = base_score + sum(contributions)

    return {
        "contributions": contributions,
        "baseScore": base_score,
        "predictedScore": predicted_score,
        "reconstructedScore": reconstructed_from_centered or reconstructed_score,
        "fidelity": fidelity,
        "runtimeMs": int((time.time() - start) * 1000),
    }


def build_career_explainability(
    *,
    predict_score: Callable[[List[float], int], float],
    feature_vector: List[float],
    feature_keys: List[str],
    labels: Dict[str, str],
    career_index: int,
    career_name: str,
    path_key: str,
    method_preference: str = "auto",
    means: List[float],
    stds: List[float],
    additional_factors: List[Dict[str, Any]],
    gap_recommendations: Dict[str, str],
    relevant_keys: List[str] | None = None,
) -> Dict[str, Any]:
    if method_preference == "auto":
        selected_method = "lime" if len(feature_keys) > 40 else "shap"
    else:
        selected_method = method_preference

    shap_report = create_empty_method_report(method="shap", career_name=career_name, path_key=path_key)
    lime_report = create_empty_method_report(method="lime", career_name=career_name, path_key=path_key)

    if selected_method == "shap":
        shap = explain_with_shap(predict_score, feature_vector, career_index, means)
        shap_report = create_method_report(
            method="shap",
            career_name=career_name,
            path_key=path_key,
            feature_vector=feature_vector,
            keys=feature_keys,
            labels=labels,
            contributions=shap["contributions"],
            base_score=shap["baseScore"],
            predicted_score=shap["predictedScore"],
            reconstructed_score=shap["reconstructedScore"],
            fidelity=shap["fidelity"],
            runtime_ms=shap["runtimeMs"],
            additional_factors=additional_factors,
            gap_recommendations=gap_recommendations,
            relevant_keys=relevant_keys,
        )
        reason = (
            "SHAP was selected because it provides additive, stable contributions for this recommendation."
            if method_preference == "shap"
            else "SHAP was selected automatically because the feature count is small enough for a stable additive explanation."
        )
    else:
        lime = explain_with_lime(predict_score, feature_vector, career_index, means, stds)
        lime_report = create_method_report(
            method="lime",
            career_name=career_name,
            path_key=path_key,
            feature_vector=feature_vector,
            keys=feature_keys,
            labels=labels,
            contributions=lime["contributions"],
            base_score=lime["baseScore"],
            predicted_score=lime["predictedScore"],
            reconstructed_score=lime["reconstructedScore"],
            fidelity=lime["fidelity"],
            runtime_ms=lime["runtimeMs"],
            additional_factors=additional_factors,
            gap_recommendations=gap_recommendations,
            relevant_keys=relevant_keys,
        )
        reason = (
            "LIME was selected because it provides a faster local explanation for this recommendation."
            if method_preference == "lime"
            else "LIME was selected automatically because the current feature set is large and a faster local explanation is preferred."
        )

    return {
        "selectedMethod": selected_method,
        "reason": reason,
        "topCareer": shap_report if selected_method == "shap" else lime_report,
        "comparison": {
            "shap": shap_report,
            "lime": lime_report,
        },
    }
